fix: invalid yield generated

maikelordaz opened the pull request.

maikelordaz committed:
chore: change function name

maikelordaz committed:
revise: fix shares needed

Leave it as an internal function to be able to track it on tenderly

maikelordaz committed:
chore: fix some current tests

maikelordaz committed:
chore: change block number

maikelordaz committed:
chore: fix test to new block

maikelordaz committed:
chore: latest ABIs

maikelordaz committed:
chore: take correct ABI

maikelordaz committed:
feat: initial test settings

maikelordaz committed:
chore: change block to january 18

maikelordaz committed:
chore: validating values reported

MohammedTaka commented on contracts/libraries/LibYieldGeneration.sol:
```diff
@@ -93,14 +93,13 @@ library LibYieldGeneration {
-
-        return ((_collateralAmount * 10 ** 18) / pricePerShare);
+        return
+            (((_collateralAmount * 10 ** 18) / _yield.totalDeposit) * _yield.totalShares) /
```
Could we get rid of the decimal multiplication and divisions by doing:

`_collateralAmount * _yield.totalShares /  _yield.totalDeposit`

maikelordaz replied:
Commit 6044c03

MohammedTaka commented on contracts/libraries/LibYieldGeneration.sol:
```diff
@@ -93,14 +93,13 @@ library LibYieldGeneration {
-    function _ethToShares(
+    function _neededShares(
         uint _collateralAmount,
         LibYieldGenerationStorage.YieldGeneration storage _yield
```
Can this not be memory instead of storage? We are not writing to _yield here.

maikelordaz replied:
No it can't because of the nested mapping, But changed only to the needed inputs on commit 89373d2

MohammedTaka reviewed (changes requested):
Logic is solid, although some optimizations can be made. I'm also missing unit tests that test the withdrawal of yield

maikelordaz committed:
feat: needed script for tests

maikelordaz committed:
feat: yield tests

maikelordaz committed:
feat: yield calculation tests

maikelordaz committed:
chore: fix tests, new block

maikelordaz committed:
chore: change file name

maikelordaz committed:
chore: new hardhat configs

maikelordaz committed:
chore: new test on CI

maikelordaz committed:
chore: update test

maikelordaz committed:
chore: refactor

maikelordaz committed:
chore: refactor

maikelordaz committed:
chore: change function mutability

maikelordaz committed:
revise: protect division by zero

MohammedTaka committed:
feat: add reimbursement solution

MohammedTaka commented:
@maikelordaz could you simulate the the reimbursement code with the inputs from the google doc on jira?

maikelordaz committed:
chore: withdrawn yield test

maikelordaz committed:
chore: block number

maikelordaz committed:
Merge branch 'main' into TT-1765-Resolve-Invalid-yield-generated

maikelordaz committed:
revise: fix decimal issue

maikelordaz commented on contracts/facets/YGFacetZaynFi.sol:
```diff
@@ -119,4 +122,110 @@ contract YGFacetZaynFi is IYGFacetZaynFi {
+
+                // Validate the amount of shares deposited
+                uint sharesAfter = IZaynVaultV2TakaDao(vaultAddress).balanceOf(termId);
+                require(neededShares == (sharesAfter - sharesBefore), "Invalid amount of shares deposited");
```
There is a little difference that fails here. For example for the first user in the inputs the neededShares are 158058370216700 and the sharesAfter - sharesBefore are 157458744451636

MohammedTaka replied:
What we can do is we should ZapIn 2 times as much eth, then zapput the difference in before and after so we have an exact amount of shares zapped in

maikelordaz commented on contracts/facets/YGFacetZaynFi.sol:
```diff
@@ -119,4 +122,110 @@ contract YGFacetZaynFi is IYGFacetZaynFi {
+                uint pricePerShare = IZaynVaultV2TakaDao(vaultAddress)
+                    .getPricePerFullShare();
+
+                uint neededEth = neededShares * pricePerShare;
```
Here I fixed the decimals with neededShares * pricePerShare / 10 ** 18

MohammedTaka replied:
Yeah forgot that, my bad. Please commit

maikelordaz committed:
chore: format

maikelordaz committed:
revise: add extra shares to avoid slippage

maikelordaz committed:
chore: fix calculation on yieldDistributionRatio

maikelordaz committed:
chore: fix tests

maikelordaz committed:
chore: update mappings

maikelordaz committed:
revise: avoid underflows

maikelordaz committed:
chore: local deployments

maikelordaz committed:
chore: clean code in test and script

MohammedTaka committed:
revise: avoid future underflows

MohammedTaka committed:
revise: avoid future underflows fix

maikelordaz committed:
chore: contracts needed to deploy upgrades on test environment

maikelordaz committed:
chore: update script

maikelordaz committed:
chore: change file name

maikelordaz committed:
feat: script to do diamondCut on tests

maikelordaz committed:
Merge branch 'TT-1765-Resolve-Invalid-yield-generated' of https://github.com/TakafulDAO/takaturn into TT-1765-Resolve-Invalid-yield-generated

MohammedTaka committed:
revise: fix & optimize rescueStuckYields

MohammedTaka committed:
feat: added restoreYieldBalance function for terms that got an imbalance due to the underflow bug

MohammedTaka committed:
chore: fix comments

maikelordaz committed:
chore: local deployments

maikelordaz committed:
feat: emit new events

maikelordaz committed:
chore: test call

maikelordaz committed:
chore: local deployment

maikelordaz committed:
chore: add new selector

maikelordaz committed:
revise: remove unneded fix. Back to normal behaviour

maikelordaz committed:
chore: change some requires for continue blocks

MohammedTaka commented on contracts/facets/YGFacetZaynFi.sol:
```diff
@@ -211,16 +218,16 @@ contract YGFacetZaynFi is IYGFacetZaynFi {
+                // Give the extra eth back to msg.sender
+                usedValue -= withdrawnExtraEth;
+
+                emit OnYieldCompensated(termId, user, neededEth);
```
The third parameter should be: neededEth - withdrawnExtraEth

MohammedTaka commented on contracts/facets/YGFacetZaynFi.sol:
```diff
@@ -235,6 +242,118 @@ contract YGFacetZaynFi is IYGFacetZaynFi {
+                ++i;
+            }
+
+            emit OnYieldTermUpdated(termId);
```
This one seems redundant, maybe add how much ETH was restored and how much was compensated. So add: withdrawnTooMuch, (neededEth - withdrawnExtraEth) & the final balanceOf from zaynfi

MohammedTaka reviewed (changes requested):
Looking good, some minor points and I think we're good to go

maikelordaz committed:
revise: fix event inputs

maikelordaz committed:
chore: fix initialization bytes

maikelordaz committed:
feat: new checks on unit tests

maikelordaz commented:
@MohammedTaka I add new checks when withdrawing collateral on the unit tests

MohammedTaka committed:
feat: Added extra function reimburseExtraYield to get rid of the surplus

MohammedTaka commented:
Added an additional function to make compensate for the additional yield that was reimbursed with a zaynfi bug.

maikelordaz committed:
chore: local ABIs

maikelordaz committed:
chore: add new selector

MohammedTaka committed:
revise: Fixed reimburseExtraYield and added more checks

MohammedTaka reviewed (commented):
We need to make sure withdrawing shares does not cause rounding errors

MohammedTaka committed:
feat: add a means to prevent rounding errors when withdrawing ETH from yield generation

MohammedTaka committed:
revise: fixed input that caused an error

MohammedTaka committed:
revise: small optimization on reimburseExtraYield

MohammedTaka committed:
revise: fix rounding error solution in _withdrawYG

elallamTaka committed:
revise(_withdrawYG): restructure to avoid underflow

if  for ex neededShares = 150,000 and sharesBalance = 10,000, then the else if would have causes an underflow

elallamTaka reviewed (approved):
BismilAllah
